nightingale_navigation/src/room_runner.py

```diff
@@ -0,0 +1,288 @@
+        self.pose_write_period = rospy.get_param("/room_runner/pose_write_period", 60)
+
+        """
+        Cached pose file has the expected form: x y theta
+        """
+        self.EXPECTED_NUM_LINES = 1
+        self.EXPECTED_NUM_VALUES = 3
+        # this is the cov matrix rviz uses when we supply an initial pose
+        self.INITAL_COV = np.reshape(
+            np.diag([0.25, 0.25, 0, 0, 0, 0.06853892]), (36,)
+        ).tolist()
+
+        if self.save_pose and not os.path.isabs(self.pose_filename):
+            rospy.logfatal(
+                f"Pose file path {self.pose_filename} is not absolute! Please provide an absolute path",
+                logger_name=self.logger_name,
+            )
+
+        if self.save_pose and os.path.exists(self.pose_filename):
+            x, y, yaw = self.read_pose()
+            self.publish_initial_pose(x, y, yaw)
+        elif self.save_pose:
+            rospy.logwarn(
+                "Pose file path does not yet exist, It will be written",
+                logger_name=self.logger_name,
+            )
+
+        self.server = SimpleActionServer(
+            "/nightingale_navigation/room_runner",
+            RoomRunnerAction,
+            self.execute_cb,
+            False,
+        )
+        self.server.register_preempt_callback(self.preempt_cb)
+        self.server.start()
+
+        move_base_action_name = "move_base"
+        self.client = SimpleActionClient(move_base_action_name, MoveBaseAction)
+        rospy.loginfo(
+            f"Waiting for {move_base_action_name} action server",
+            logger_name=self.logger_name,
+        )
+        if self.client.wait_for_server():
+            rospy.loginfo(
+                f"Found the {move_base_action_name} action server",
+                logger_name=self.logger_name,
+            )
+        else:
+            rospy.logfatal(
+                f"Failed to find the {move_base_action_name} action server",
+                logger_name=self.logger_name,
+            )
+
+        self.timer = (
+            rospy.Timer(rospy.Duration(self.pose_write_period), self.write_pose)
+            if self.save_pose
+            else None
+        )
+
+        self.home = "home"
+
+    def execute_cb(self, goal):
+        rospy.loginfo(f"Received room_runner action call", logger_name=self.logger_name)
+        door_goal_pose = self.call_door_lookup(goal.patient_name)
+        msg, fb_cb, a_cb = self.generate_goal_info(door_goal_pose, "navigating to door")
+
+        self.client.send_goal(msg, feedback_cb=fb_cb, active_cb=a_cb)
+        self.client.wait_for_result()
+
+        # move base result is empty so no need to check
+        result = RoomRunnerResult()
+        if self.client.get_state() == GoalStatus.SUCCEEDED:
+            rospy.loginfo(
+                "Nav to Room Completed Successfully :)", logger_name=self.logger_name
+            )
+        elif self.client.get_state() == GoalStatus.PREEMPTED:
```

Comment:
When a call is preempted what happens? Right now does the robot just stop in place and wait?? Would it be useful to have the robot give up and go home or is it safest to stop and wait for a technician/someone to come fix the robot?

Reply:
When an action is preempted, the robot cancels the move base action and stops moving. It is up to the caller to decide what to do next. If the caller would like the robot to go home, they should call the actino again with 'home' as the goal